### DISCUSSION TITLE
Frontend Events?

### DISCUSSION THREAD

**Discussion by meichthys** [TriliumNext/Trilium#3926] [Q&A]
Currently the documentation seems to indicate that all existing triggers must be tied to a backend script (not frontend).

I'm trying to create a script that runs when the frontend tries to sync a note change (like `runOnNoteChange` but on the frontend).
My idea is to update the note label with the ip address of the client that made the change (it seems like only the frontend could determine it's own ip).

Would this be possible with the existing events/functionality of Trilium?

**Comment by meichthys** [comment]
For example, this frontend script does what i want, but so far i can only call it by manually executing the script (I don't see a way to call this when the note is changed):

```
/**
 * Get the user IP throught the webkitRTCPeerConnection
 * @param onNewIP {Function} listener function to expose the IP locally
 * @return undefined
 */
function getUserIP(onNewIP) { //  onNewIp - your listener function for new IPs
    //compatibility for firefox and chrome
    var myPeerConnection = window.RTCPeerConnection || window.mozRTCPeerConnection || window.webkitRTCPeerConnection;
    var pc = new myPeerConnection({
        iceServers: []
    }),
    noop = function() {},
    localIPs = {},
    ipRegex = /(192(\.[0-9]{1,3}){3}|[a-f0-9]{1,4}(:[a-f0-9]{1,4}){7})/g,
    key;

    function iterateIP(ip) {
        console.log(ip);
        if (!localIPs[ip]) onNewIP(ip);
        localIPs[ip] = true;
    }

     //create a bogus data channel
    pc.createDataChannel("");

    // create offer and set local description
    pc.createOffer().then(function(sdp) {
        sdp.sdp.split('\n').forEach(function(line) {
            if (line.indexOf('candidate') < 0) return;
            line.match(ipRegex).forEach(iterateIP);
        });
        
        pc.setLocalDescription(sdp, noop, noop);
    }).catch(function(reason) {
        // An error occurred, so handle the failure to connect
    });

    //listen for candidate events
    pc.onicecandidate = function(ice) {
        if (!ice || !ice.candidate || !ice.candidate.candidate || !ice.candidate.candidate.match(ipRegex)) return;
        ice.candidate.candidate.match(ipRegex).forEach(iterateIP);
    };
}
   
// Usage

getUserIP(function(ip){
    api.getActiveContextNote().getAttribute("label", "lastEditIp").value = ip
});
```

**Comment by zadam** [comment]
It can be done using [custom widgets](https://github.com/zadam/trilium/wiki/Custom-widget).

This is not really documented / part of the script API, but widgets receive events: 

https://github.com/zadam/trilium/blob/ad2a7a2da36bf3c0c9bb9b4eaca68f8b3853632f/src/public/app/widgets/mermaid.js#L110-L114

You can find many other examples in the codebase of handling the `entitiesReloadedEvent`.

Given you don't need UI, just render an empty `<span></span>` or something like that. It's enough that the widget is included into the component tree so it receives events.

**Reply by meichthys** [reply]
Thanks. I'll take a look!
